### PR TITLE
MNG-6685: convert maven-model-builder to JSR 330

### PR DESCRIPTION
**Pull request by josephw**
Further conversion to JSR 330 annotations, following [Plexus Migration](https://wiki.eclipse.org/Sisu/PlexusMigration), for `maven-model-builder`, with some test changes to adjust.

In tests, use of `PlexusTestCase` for trivial wiring is reduced, to directly construct the classes under test. Tests that do more significant wiring are modified to use `PlexusConstants.SCANNING_ON` so JSR 330-annotated classes no longer included in component indexes are picked up.

Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [X] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MNG) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [X] Each commit in the pull request should have a meaningful subject line and body.
 - [X] Format the pull request title like `[MNG-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MNG-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [X] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [X] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [X] You have run the [Core IT][core-its] successfully.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [X] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

[core-its]: https://maven.apache.org/core-its/core-it-suite/

### CODE REVIEW

File: maven-core/src/test/java/org/apache/maven/project/artifact/MavenMetadataSourceTest.java [hunk 1]
```diff
@@ -21,7 +21,9 @@
 
 import org.apache.maven.repository.RepositorySystem;
 import org.codehaus.plexus.PlexusTestCase;
+import org.junit.Ignore;
 
+@Ignore
```
Comment (michael-o) [inline maven-core/src/test/java/org/apache/maven/project/artifact/MavenMetadataSourceTest.java]: Why is this one skipped?

File: maven-core/src/test/java/org/apache/maven/settings/PomConstructionWithSettingsTest.java [hunk 2]
```diff
@@ -60,7 +60,7 @@ protected void customizeContainerConfiguration( ContainerConfiguration container
     {
         super.customizeContainerConfiguration( containerConfiguration );
         containerConfiguration.setAutoWiring( true );
-        containerConfiguration.setClassPathScanning( PlexusConstants.SCANNING_INDEX );
+        containerConfiguration.setClassPathScanning( PlexusConstants.SCANNING_ON );
```
Comment (michael-o) [inline maven-core/src/test/java/org/apache/maven/settings/PomConstructionWithSettingsTest.java]: Why does this change? As far as I remember, Maven doesn't do classpath scanning, it relies on descriptors. @rfscholte correct me if I am wrong.

File: maven-embedder/src/test/java/org/apache/maven/cli/CLIManagerDocumentationTest.java [hunk 3]
```diff
@@ -100,7 +101,7 @@ public String getOptionsAsHtml()
     public void testOptionsAsHtml()
         throws IOException
     {
-        File options = getTestFile( "target/test-classes/options.html" );
+        File options = new File( "target/test-classes/options.html" );
```
Comment (michael-o) [inline maven-embedder/src/test/java/org/apache/maven/cli/CLIManagerDocumentationTest.java]: Why this change?

File: maven-model-builder/src/main/java/org/apache/maven/model/interpolation/StringSearchModelInterpolator.java [hunk 4]
```diff
@@ -45,10 +44,14 @@
...
 
+import javax.inject.Named;
+import javax.inject.Singleton;
+
 /**
  * StringSearchModelInterpolator
  */
-@Component( role = ModelInterpolator.class )
+@Named
+@Singleton
```
Comment (michael-o) [inline maven-model-builder/src/main/java/org/apache/maven/model/interpolation/StringSearchModelInterpolator.java]: Will this properly determine the role?

File: maven-model-builder/src/main/java/org/apache/maven/model/io/DefaultModelReader.java [hunk 5]
```diff
@@ -27,21 +27,26 @@
...
 
 /**
  * Handles deserialization of a model from some kind of textual format like XML.
  *
  * @author Benjamin Bentmann
  */
-@Component( role = ModelReader.class )
+@Named
+@Singleton
+@Typed( ModelReader.class )
```
Comment (michael-o) [inline maven-model-builder/src/main/java/org/apache/maven/model/io/DefaultModelReader.java]: Why is the `@Typed` here?

File: maven-resolver-provider/src/test/java/org/apache/maven/repository/internal/AbstractRepositoryTestCase.java [hunk 6]
```diff
@@ -44,7 +44,7 @@ protected void customizeContainerConfiguration( ContainerConfiguration container
     {
         super.customizeContainerConfiguration( containerConfiguration );
         containerConfiguration.setAutoWiring( true );
-        containerConfiguration.setClassPathScanning( PlexusConstants.SCANNING_INDEX );
+        containerConfiguration.setClassPathScanning( PlexusConstants.SCANNING_ON );
```
Comment (michael-o) [inline maven-resolver-provider/src/test/java/org/apache/maven/repository/internal/AbstractRepositoryTestCase.java]: Same here...

File: maven-core/src/test/java/org/apache/maven/project/artifact/MavenMetadataSourceTest.java (hunk 1 above)
Comment (josephw) [reply maven-core/src/test/java/org/apache/maven/project/artifact/MavenMetadataSourceTest.java]: All test methods are commented out since a73c5abc9c584 and b8c6c01864e. This change prevents the `setUp` from causing a problem, without trying to fix a test that doesn't currently run, but stops short of deleting the test.

The commenting-out is from 2009; is this a good point to drop the class?

File: maven-core/src/test/java/org/apache/maven/settings/PomConstructionWithSettingsTest.java (hunk 2 above)
Comment (josephw) [reply maven-core/src/test/java/org/apache/maven/settings/PomConstructionWithSettingsTest.java]: This change is necessary for the JSR 330 classes to get picked up in `PlexusTestCase`. Maven runs after these changes (and `maven-integration-testing` passes), so it looks like something's picking them up without descriptors (something in [`MavenCli`](https://github.com/apache/maven/blob/edf61684c834090561954ea32d90d917e69c7b5e/maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java)?).

File: maven-embedder/src/test/java/org/apache/maven/cli/CLIManagerDocumentationTest.java (hunk 3 above)
Comment (josephw) [reply maven-embedder/src/test/java/org/apache/maven/cli/CLIManagerDocumentationTest.java]: These classes no longer extend `PlexusTestCase`, so they lose the `getTestFile` utility method, that [effectively does just this](https://github.com/codehaus-plexus/plexus-containers/blob/dfb1d7132c7f52122a6e18da8a9fb86e7592dee2/plexus-container-default/src/main/java/org/codehaus/plexus/PlexusTestCase.java#L229).

File: maven-model-builder/src/main/java/org/apache/maven/model/interpolation/StringSearchModelInterpolator.java (hunk 4 above)
Comment (josephw) [reply maven-model-builder/src/main/java/org/apache/maven/model/interpolation/StringSearchModelInterpolator.java]: Yes; it's also the [advice in Sisu/PlexusMigration](https://wiki.eclipse.org/Sisu/PlexusMigration).

File: maven-model-builder/src/main/java/org/apache/maven/model/io/DefaultModelReader.java (hunk 5 above)
Comment (josephw) [reply maven-model-builder/src/main/java/org/apache/maven/model/io/DefaultModelReader.java]: This one case is ambiguous; without it, `org.apache.maven.repository.internal.RepositorySystemTest` fails with `StackOverflow` when `DefaultModelProcessor` gets itself injected as its own `ModelReader` (because `ModelProcessor` extends `ModelReader`).

There may be a better approach, but that's why it's here.

File: maven-embedder/src/test/java/org/apache/maven/cli/CLIManagerDocumentationTest.java (hunk 3 above)
Comment (michael-o) [reply maven-embedder/src/test/java/org/apache/maven/cli/CLIManagerDocumentationTest.java]: Looking at `getBasedir()` won't this fail is `-Dbasedir` is set and you don't read it?

Comment (josephw) [reply maven-embedder/src/test/java/org/apache/maven/cli/CLIManagerDocumentationTest.java]: In this case, the test file is stored in that specific location, so setting `-Dbasedir` to anything else would cause the existing test to fail! It doesn't look like `basedir`'s being set anywhere for these tests, so that behaviour may not need to be carried over.

File: maven-model-builder/src/main/java/org/apache/maven/model/interpolation/StringSearchModelInterpolator.java (hunk 4 above)
Comment (michael-o) [reply maven-model-builder/src/main/java/org/apache/maven/model/interpolation/StringSearchModelInterpolator.java]: How, if I look at the Type Override section with `extends` it explicitly references the interface name.

File: maven-core/src/test/java/org/apache/maven/settings/PomConstructionWithSettingsTest.java (hunk 2 above)
Comment (michael-o) [reply maven-core/src/test/java/org/apache/maven/settings/PomConstructionWithSettingsTest.java]: Does it fail if you don't?

File: maven-model-builder/src/main/java/org/apache/maven/model/io/DefaultModelReader.java (hunk 5 above)
Comment (michael-o) [reply maven-model-builder/src/main/java/org/apache/maven/model/io/DefaultModelReader.java]: I understand the fact about the recursion, but I don't understand why *this* class requires it. Isn't the type here straight forward? `DefaultModelProcessor` is aready annotated with `@Component( role = ModelProcessor.class )` how can it still be injected as `ModelReader`?

File: maven-core/src/test/java/org/apache/maven/settings/PomConstructionWithSettingsTest.java (hunk 2 above)
Comment (josephw) [reply maven-core/src/test/java/org/apache/maven/settings/PomConstructionWithSettingsTest.java]: Odd. Checking again, it passes without these changes.
It could have been an IDE issue; either way, I'll roll them back.

File: maven-model-builder/src/main/java/org/apache/maven/model/interpolation/StringSearchModelInterpolator.java (hunk 4 above)
Comment (josephw) [reply maven-model-builder/src/main/java/org/apache/maven/model/interpolation/StringSearchModelInterpolator.java]: From that document:
> though in some rare cases an explicit type is required.
>...
> `extends SomeSupportClassHardToGuessTypeFrom`

My impression from the code (is it [this part](https://github.com/eclipse/sisu.inject/blob/f4d69f725f23cc3b7dabedc5ee8d0fbec665add8/org.eclipse.sisu.inject/src/org/eclipse/sisu/space/QualifiedTypeBinder.java#L407)?) is that a simple type hierarchy with interfaces is fine, even if it goes via another class. That note is about more complex cases than this.

As a test, if I add:

    @Typed(AbstractStringBasedModelInterpolator.class)

to intentionally mis-bind, the test fails (with no `ModelInterpolator`). So I'm confident it's being properly determined here.

File: maven-model-builder/src/main/java/org/apache/maven/model/io/DefaultModelReader.java (hunk 5 above)
Comment (josephw) [reply maven-model-builder/src/main/java/org/apache/maven/model/io/DefaultModelReader.java]: You're right, this is just repeating the (obvious) type that would be inferred. Testing again, it doesn't seem to be necessary.

However, on `DefaultModelProcessor`, with `Component` now replaced by `@Named @Singleton`, it's made available as *all* its implemented interfaces. I've added a `@Typed` to restrict that to `ModelProcessor`.

I'm not able to explain why I no longer see recursive wiring, but that `@Typed` should prevent it.

**Comment by michael-o** [comment]
I have pushed a branch now, but had to reintroduce the change with the scanning for `DefaultProfileManagerTest` because of injection errors.

**Comment by qoomon** [comment]
I've build a core extension https://github.com/qoomon/maven-git-versioning-extension
however it stoped working after maven version 3.6.0
My replacement for ModelProcessor https://github.com/qoomon/maven-git-versioning-extension/blob/master/src/main/java/me/qoomon/maven/gitversioning/ModelProcessor.java
will still be initialised however it is not used as ModelProcessor anymore.
Any clues? Kind regards

**Comment by michael-o** [comment]
Wrong place. Ask on the mailing list.

**Comment by qoomon** [comment]
How to send to which mailing list?

**Comment by soloturn** [comment]
@michael-o - https://github.com/apache/maven/commit/07ab962c85950b034be3216996900920c0204c3a

**Comment by jira-importer** [comment]
Resolve #8636
